containers/src/Kokkos_DynamicView.hpp

```diff
@@ -53,36 +53,184 @@
+                               (sizeof(value_type*) * (m_chunk_max + 2)));
+    }
+  }
+
+  void destroy_shared_allocation() { execute(); }
+
+  std::string m_label;
+  value_type** m_chunks = nullptr;
+  value_type** m_linked = nullptr;
+  unsigned m_chunk_max;
+  unsigned m_chunk_size;
 };
-#endif
-#ifdef KOKKOS_ENABLE_SYCL
-template <>
-struct ChunkArraySpace<Kokkos::Experimental::SYCLDeviceUSMSpace> {
-  using memory_space = typename Kokkos::Experimental::SYCLSharedUSMSpace;
+
+template <typename MemorySpace, typename ValueType>
+struct SmartMemoryAccessor {
+  using pointer_type = ValueType*;
+  using track_type   = Kokkos::Impl::SharedAllocationTracker;
+
+  SmartMemoryAccessor()                           = default;
+  SmartMemoryAccessor(SmartMemoryAccessor const&) = default;
+  SmartMemoryAccessor(SmartMemoryAccessor&&)      = default;
+  SmartMemoryAccessor& operator=(SmartMemoryAccessor&&) = default;
+  SmartMemoryAccessor& operator=(const SmartMemoryAccessor&) = default;
+
+  template <typename Space, typename Value>
+  friend struct SmartMemoryAccessor;
+
+  template <typename Space, typename Value>
+  inline SmartMemoryAccessor(const SmartMemoryAccessor<Space, Value>& rhs)
+      : m_valid(rhs.m_valid),
+        m_chunk_max(rhs.m_chunk_max),
+        m_chunks((ValueType**)(rhs.m_chunks)),
+        m_track(rhs.m_track),
+        m_chunk_size(rhs.m_chunk_size) {
+    static_assert(
+        Kokkos::Impl::MemorySpaceAccess<MemorySpace, Space>::assignable,
+        "Incompatible SmartMemoryAccessor copy construction");
+  }
+
+  SmartMemoryAccessor(const unsigned arg_chunk_max,
+                      const unsigned arg_chunk_size)
+      : m_chunk_max(arg_chunk_max), m_chunk_size(arg_chunk_size) {}
+
+ private:
+  struct ACCESSIBLE_TAG {};
+  struct INACCESSIBLE_TAG {};
+
+  SmartMemoryAccessor(ACCESSIBLE_TAG, pointer_type* arg_chunks,
+                      const unsigned arg_chunk_max)
+      : m_valid(true), m_chunk_max(arg_chunk_max), m_chunks(arg_chunks) {}
+
+  SmartMemoryAccessor(INACCESSIBLE_TAG, const unsigned arg_chunk_max,
+                      const unsigned arg_chunk_size)
+      : m_chunk_max(arg_chunk_max), m_chunk_size(arg_chunk_size) {}
+
+ public:
+  template <typename Space>
+  static constexpr bool is_accessible_from =
+      Kokkos::Impl::MemorySpaceAccess<MemorySpace, Space>::accessible;
+
+  template <typename Space>
+  static SmartMemoryAccessor<Space, ValueType> create_mirror(
+      SmartMemoryAccessor<MemorySpace, ValueType> const& other,
+      typename std::enable_if<is_accessible_from<Space>>::type* = nullptr) {
+    return SmartMemoryAccessor<Space, ValueType>{
+        ACCESSIBLE_TAG{}, other.m_chunks, other.m_chunk_max};
+  }
+
+  template <typename Space>
+  static SmartMemoryAccessor<Space, ValueType> create_mirror(
+      SmartMemoryAccessor<MemorySpace, ValueType> const& other,
+      typename std::enable_if<not is_accessible_from<Space>>::type* = nullptr) {
```

Comment:
make that ! not "not" msvc doesn't like it without special flags (mostly for C conformity I think)